Return resulting `URI` from commands that save the active editor (fix #178713)

bpasero opened the pull request:
Fixes https://github.com/microsoft/vscode/issues/178713, https://github.com/microsoft/vscode/issues/124117, https://github.com/microsoft/vscode/issues/22040

@jrieken mainly 👁️ for the new API commands, I feel `vscode.save` and `vscode.saveAs` are better then asking extensions to use our internal commands, especially since that allows us to explicitly have a typed return type. I decided to not just return `vscode.Uri` but `{ uri: vscode.Uri }` because we might later decide to add the view type to the result as well. 

Conceptually saving an editor can result in a different editor opening because it is possible that saving an untitled editor changes the view type (e.g. saving an untitled file as notebook reopens as notebook).

jrieken commented:
Not a fan of an having an API command to save the "active editor" because it's really fishy that is and will always change under you. Something more reliable would be a save-command that takes an URI of the resource that you want to save or save-as

bpasero commented:
@jrieken changed to require a `URI` for both commands and updated the tests. 

Btw this will also address https://github.com/microsoft/vscode/issues/22040 which asks for a way to force a save to disk even when the file is not dirty.

bpasero commented:
@jrieken as discussed, changed to new proposed API `workspace.save(uri): URI` and `workspace.saveAs(uri): URI` and added tests. This goes along nice with the existing `workspace.saveAll()`.

jrieken commented on src/vs/workbench/api/common/extHostWorkspace.ts:
```diff
@@ -557,6 +557,18 @@ export class ExtHostWorkspace implements ExtHostWorkspaceShape, IExtHostWorkspac
+	async saveAs(uri: URI): Promise<URI | undefined> {
+		const result = await this._proxy.$saveAs(uri);
+
+		return URI.revive(result);
```
please make sure to enforce both proposals

jrieken replied:
🤦 didn't see that. checking in extHost.api is best

jrieken commented on src/vs/workbench/api/browser/mainThreadWorkspace.ts:
```diff
@@ -201,8 +203,34 @@ export class MainThreadWorkspace implements MainThreadWorkspaceShape {
+		return firstOrDefault(this.saveResultToUris(result));
+	}
+
+	async $saveAs(uriComponents: UriComponents): Promise<UriComponents | undefined> {
```
nit: only have `$save` but with `$save(uri, saveAs)`

jrieken commented on src/vs/workbench/api/browser/mainThreadWorkspace.ts:
```diff
@@ -201,8 +203,34 @@ export class MainThreadWorkspace implements MainThreadWorkspaceShape {
+		return firstOrDefault(this.saveResultToUris(result));
+	}
+
+	private saveResultToUris(result: ISaveEditorsResult): URI[] {
```
`_` ;-)

jrieken commented on src/vscode-dts/vscode.proposed.saveEditor.d.ts:
```diff
@@ -0,0 +1,30 @@
+	export namespace workspace {
+
+		/**
+		 * Saves the editor identified by the given resource and returns the resulting resource or `undefined`
```
maybe make it more clear that save only works when a respective editor is open

bpasero commented:
PR review feedback will be addressed in https://github.com/microsoft/vscode/pull/180476
